client-tez/src/main/java/org/apache/tez/runtime/library/common/shuffle/orderedgrouped/RssTezBypassWriter.java

```diff
@@ -0,0 +1,94 @@
+import org.apache.tez.runtime.library.common.shuffle.DiskFetchedInput;
+import org.apache.tez.runtime.library.common.shuffle.FetchedInput;
+import org.apache.tez.runtime.library.common.shuffle.MemoryFetchedInput;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+// In Tez shuffle, MapOutput encapsulates the logic to fetch map task's output data via http.
+// So, in RSS, we should bypass this logic, and directly write data to MapOutput.
+public class RssTezBypassWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(RssTezBypassWriter.class);
+  private static final byte[] HEADER = new byte[] { (byte) 'T', (byte) 'I', (byte) 'F', (byte) 0};
+
+  public static void write(MapOutput mapOutput, byte[] buffer) {
+    // Write and commit uncompressed data to MapOutput.
+    // In the majority of cases, merger allocates memory to accept data,
+    // but when data size exceeds the threshold, merger can also allocate disk.
+    // So, we should consider the two situations, respectively.
+    if (mapOutput.getType() == MapOutput.Type.MEMORY) {
+      byte[] memory = mapOutput.getMemory();
+      System.arraycopy(buffer, 0, memory, 0, buffer.length);
+    } else if (mapOutput.getType() == MapOutput.Type.DISK) {
+      // RSS leverages its own compression, it is incompatible with hadoop's disk file compression.
+      // So we should disable this situation.
+      throw new IllegalStateException("RSS does not support OnDiskMapOutput as shuffle ouput,"
+              + " try to reduce mapreduce.reduce.shuffle.memory.limit.percent");
+    } else {
+      throw new IllegalStateException("Merger reserve unknown type of MapOutput: "
+              + mapOutput.getClass().getCanonicalName());
+    }
+  }
+
+
+  public static void write(final FetchedInput mapOutput, byte[] buffer) throws IOException {
+    // Write and commit uncompressed data to MapOutput.
+    // In the majority of cases, merger allocates memory to accept data,
+    // but when data size exceeds the threshold, merger can also allocate disk.
+    // So, we should consider the two situations, respectively.
+    if (mapOutput.getType() == FetchedInput.Type.MEMORY) {
+      byte[] memory = ((MemoryFetchedInput) mapOutput).getBytes();
+      System.arraycopy(buffer, 0, memory, 0, buffer.length);
+    } else if (mapOutput.getType() == FetchedInput.Type.DISK) {
+      OutputStream output = ((DiskFetchedInput) mapOutput).getOutputStream();
+      output.write(HEADER);
+      output.write(buffer);
+      output.write(calcChecksum(buffer));
+      output.flush();
+      output.close();
+    } else {
+      throw new IllegalStateException("Merger reserve unknown type of MapOutput: "
+          + mapOutput.getClass().getCanonicalName());
+    }
+  }
+
+  @VisibleForTesting
+  protected static byte[] calcChecksum(final byte[] buffer) throws IOException {
+    DataChecksum sum = DataChecksum.newDataChecksum(DataChecksum.Type.CRC32, Integer.MAX_VALUE);
```

Comment:
Why do we need the privillege `protected`?

Reply:
ok, we removed this method  and use org.apache.uniffle.common.util.ChecksumUtils#getCrc32